File: Leetcode/744 - Find Smallest Letter Greater Than Target/solution.ts
function nextGreatestLetter(letters: string[], target: string): string {
    let firstIdx: number = 0;
    let lastIdx: number = letters.length - 1;

    while (firstIdx <= lastIdx) {
        const midIdx: number = Math.floor((firstIdx + lastIdx) / 2);
        if (letters[midIdx] <= target) {
            firstIdx = midIdx + 1;
        } else {
            lastIdx = midIdx - 1;
        }
    }

    // If firstIdx goes past the array, wrap around
    return letters[firstIdx % letters.length];
}

File: Leetcode/744 - Find Smallest Letter Greater Than Target/solution.py
class Solution:
    def nextGreatestLetter(self, letters: List[str], target: str) -> str:
        firstIdx: int = 0
        lastIdx: int = len(letters) - 1

        while (firstIdx <= lastIdx):
            MIDIDX: int = (firstIdx + lastIdx) // 2

            if (letters[MIDIDX] <= target):
                firstIdx = MIDIDX + 1
            else:
                lastIdx = MIDIDX - 1

        # If left goes past the array, wrap around
        return letters[firstIdx % len(letters)]
        